pkg/ebpf/c/tracee.bpf.c

```diff
@@ -499,56 +499,76 @@ int tracepoint__sched__sched_process_fork(struct bpf_raw_tracepoint_args *ctx)
     if (!init_program_data(&p, ctx))
         return 0;
 
-    // Note: we don't place should_trace() here, so we can keep track of the cgroups in the system
-    struct task_struct *parent = (struct task_struct *) ctx->args[0];
+    // no should_trace() here so we can keep track of the cgroups in the system
+
     struct task_struct *child = (struct task_struct *) ctx->args[1];
+    struct task_struct *leader = get_leader_task(child);
+    struct task_struct *parent = get_leader_task(get_parent_task(leader));
```

Comment:
I know and that was the intent. If the proccess tree is built using the "parent" semantics pointing to a real process, why sched_process_XXX events should be different ? Specially considering we might (will) attach "artifacts" (used files, used sockets, etc) to a "real process" (and not per thread).

Reply:
I see the problem here, and I have to agree here with @yanivagman.
Someone using the `sched_process_fork` might be interested in the real parent task, and not necessarily the leader or the process parent.
This is something worth discussing.
Currently, even getting the thread from the process tree is impossible because of lacking of information in the general event (the thread start time is since epoch and not since start time). It will require us to add at least something to query the tree on the thread to the event structure. Maybe we should consider adding the process information too? This might ease this problem.
However, giving the thread, process and parent process hash is a bit extreme maybe (though each hash is only 32bit), and start time will be even more so (especially when the start time of the task since boot is probably irrelevant to the user except for the process tree).
So I am for either giving thread, process and parent process hashes in each event, or giving only the thread and process while adding argument for the parent process to the `sched_process_fork` event.

Reply:
> Currently, even getting the thread from the process tree is impossible because of lacking of information in the general event (the thread start time is since epoch and not since start time). It will require us to add at least something to query the tree on the thread to the event structure. 

In each event you have "Entity ID (tid) + Starttime (not epoch) Hash". You can find that event's task (process or thread) using the entity ID, no ?

I'll review https://github.com/aquasecurity/tracee/pull/3503 to check what you mean.

Reply:
> You are right that you get the task entity id. However, it is nit exported using json, so it is not applicable in analyze mode.

OHHHHH true true true, shit.

Reply:
![image](https://github.com/aquasecurity/tracee/assets/7395852/0f4f58ab-8c87-4b57-b2a7-24ab8b4aaa36)

Reply:
Parent entity id is going to be part of the new event structure anyway: https://github.com/aquasecurity/tracee/issues/2870

Reply:
Yet, I think we should keep the sematics of having the real parent (be it a thread or not)

Reply:
The process tree REALLY needs 2 things that the regular sched_process_fork() does not provide:

- leader info
- parent (a real process) info

Since we don't want to touch the existing fork event (its parent) nor add fields (unless needed by something else) I created a small logic of only submitting "leader" and "up_parent" IF a config flag exists (whether proctree is using the regular events as source). 

![image](https://github.com/aquasecurity/tracee/assets/7395852/8f505712-5ae8-4874-853d-3a5f4e31406d)

And then, in userland, after processed by proc tree, I remove these extra arguments (because even if the proc tree is not enabled they would be displayed as "nil", and we dont want them displayed at all):

![image](https://github.com/aquasecurity/tracee/assets/7395852/ad85e9ff-4b65-48c5-a198-e703750e2aaa)

![image](https://github.com/aquasecurity/tracee/assets/7395852/0c5743da-b2c9-43de-913c-013d9f8760c9)

This "solves" both problems. And, yes, using regular pipeline will be needed as observed by @AlonZivony in his recent tests. Something to discuss later (when the pipeline is small, specially for tests, there are good chances that the fork signal event would arrive after the "exec" regular event, for example).